### PR TITLE
CI: E2E

### PR DESCRIPTION
**Pull request by tommyp1ckles**
Adds E2E conformance testing to repo, provides matrix of tests to provide reasonable coverage on various common features and configuration types.

**Cilium Versions:**
* v1.15
* v1.14

**Talos Versions:**
* v1.6.2

--- 

**Features:**

* TunnelMode: VXLAN, Geneve
* Encryption: IPSec, Wireguard
* BPF Masquerading: Enabled, Disabled
* Kube Proxy Replacement: Enabled, Disabled
* IPAM Mode: ClusterPool, Kubernetes

### CODE REVIEW

File: .github/workflows/conformance.yml [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+            encryption-enabled: false
+            encryption-type: ipsec
+            tunnel-mode: vxlan
+            nodeport: true
+
+          - name: 'v1.15 w wireguard'
+            cilium-version: "v1.15.1"
+            talos-version: 'v1.6.2'
+            kube-proxy: true 
+            kube-proxy-replacement: "false"
```
Comment (PhilipSchmid) [inline .github/workflows/conformance.yml:47]: Can we change this to KPR=true? KPR is our recommended Talos setup & the Cilium 1.14 test also runs with KPR.

File: .github/workflows/conformance.yml [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+            socketlb: false 
+            bpf-masquerade: false
```
Comment (PhilipSchmid) [inline .github/workflows/conformance.yml:49]: Same here. Can we enable this to align it with our recommendation and 1.14 testing?

File: .github/workflows/conformance.yml [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+            cilium-version: "1.14.7"
+            talos-version: 'v1.6.2'
+            kube-proxy: false
+            kube-proxy-replacement: "true"
+            socketlb: false 
+            bpf-masquerade: true
+            ipam-mode: 'kubernetes'
+            ipv4: true
+            ipv6: false
+            encryption-enabled: false
```
Comment (PhilipSchmid) [inline .github/workflows/conformance.yml:159]:
```suggestion
            encryption-enabled: true
```

As that's the "v1.14 w wireguard" test case

File: .github/workflows/conformance.yml [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+            encryption-type: wireguard
+            tunnel-mode: vxlan
+            nodeport: false
```
Comment (PhilipSchmid) [inline .github/workflows/conformance.yml:162]:
```suggestion
            nodeport: true
```

Would be required for `bpf-masquerade: true`, no? https://docs.cilium.io/en/stable/network/concepts/masquerading/#ebpf-based

File: .github/workflows/conformance.yml [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+            kube-proxy-replacement: "false"
+            socketlb: false 
+            bpf-masquerade: false
+            ipam-mode: 'kubernetes'
+            ipv4: true
+            ipv6: false
+            encryption-enabled: true
+            encryption-type: wireguard
+            tunnel-mode: vxlan
+            nodeport: false
```
Comment (PhilipSchmid) [inline .github/workflows/conformance.yml:56]:
```suggestion
            nodeport: true
```

Would be required for `bpf-masquerade: true`, no? https://docs.cilium.io/en/stable/network/concepts/masquerading/#ebpf-based

File: .github/workflows/conformance.yml [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+            kube-proxy: true
+            kube-proxy-replacement: "false"
+            socketlb: true
+            bpf-masquerade: true
+            ipam-mode: 'kubernetes'
+            ipv4: true
+            ipv6: false
+            encryption-enabled: false
+            tunnel-mode: vxlan
+            nodeport: false
```
Comment (PhilipSchmid) [inline .github/workflows/conformance.yml:85]:
```suggestion
            nodeport: true
```

Would be required for `bpf-masquerade: true`, no? https://docs.cilium.io/en/stable/network/concepts/masquerading/#ebpf-based

File: test/conformance/tests/cilium-connectivity-test.sh [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+export KUBECONFIG=$(terraform output --raw path_to_kubeconfig_file)
+
+# Run the connectivity tests.
+kubectl -n kube-system port-forward svc/hubble-relay 4245:80 &
+# NS precreation is required because of https://www.talos.dev/v1.5/kubernetes-guides/configuration/pod-security/
+kubectl create ns ${ns}
+kubectl label ns ${ns} pod-security.kubernetes.io/enforce=privileged
+kubectl label ns ${ns} pod-security.kubernetes.io/warn=privileged
+PID=$!
+set +e
```
Comment (PhilipSchmid) [inline test/conformance/tests/cilium-connectivity-test.sh:17]:
```suggestion
```

We need to remove this; otherwise, a `cilium connectivity test` non-zero exit code won't be caught.

File: test/conformance/tests/cilium-connectivity-test.sh [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+cilium connectivity test --test-namespace=${ns}
+set -e
```
Comment (PhilipSchmid) [inline test/conformance/tests/cilium-connectivity-test.sh:19]:
```suggestion
```

We need to remove this; otherwise, a `cilium connectivity test` non-zero exit code won't be caught.

File: test/conformance/values.yaml [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+debug:
+  enabled: true
+ 
+# Configure unique cluster name & ID (used for ClusterMesh in the future)
+cluster:
+  id: 1
+
+# Routing/encapsulation mode
+tunnelProtocol: vxlan
+tunnel: vxlan
```
Comment (PhilipSchmid) [inline test/conformance/values.yaml:27]:
```suggestion
```
Let's remove this as it's redundant with the `tunnelProtocol` & `routingMode` combo.

File: .github/workflows/conformance.yml [hunk 10]
```diff
@@ -0,0 +1,292 @@
+name: Talos Conformance
+on:
+  pull_request_target:
+    types:
+      - opened
+      - synchronize
+      - reopened
+  schedule:
+    # Run weekly.
+    - cron: '* * * * 1'
```
Comment (PhilipSchmid) [inline .github/workflows/conformance.yml:10]: Shouldn't this be something like `0 0 * * 1`? Otherwise it's "At **every** minute on Monday."

File: .github/workflows/conformance.yml
Comment (PhilipSchmid) [inline .github/workflows/conformance.yml:1]: How do we get notified in case a weekly CI run fails?

File: .github/workflows/conformance.yml [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+    strategy:
+      fail-fast: false
+      max-parallel: 4
+      matrix:
+        include:
+          # --- Cilium v1.15 ---
+
+          - name: 'v1.15'
+            cilium-version: "v1.15.1"
+            talos-version: 'v1.6.2'
```
Comment (PhilipSchmid) [inline .github/workflows/conformance.yml:30]: How about testing multiple Talos versions? E.g., the two latest minor versions.

Comment (PhilipSchmid) [reply .github/workflows/conformance.yml:30]: Talos mainly only supports the latest minor release + critical fixes for n-1: https://www.talos.dev/v1.6/introduction/support-matrix/

Comment (tommyp1ckles) [reply .github/workflows/conformance.yml:30]: Makes sense, I'm thinking we're going to want to maybe have two sets of tests... a PR that's just most recent and a weekly one which will do a full matrix of:

(n...n-2 cilium versions) X (n...n-1 talos versions) X (config tuples) = 3 * 2 * 6 = 36

Which is probably too heavy for a PR workflow.

File: .github/workflows/conformance.yml
Comment (PhilipSchmid) [reply .github/workflows/conformance.yml:1]: As agreed on Slack, via Slack 😄.

**Review by PhilipSchmid — APPROVED** [review]
LGMT, thanks!
